mitodl/open-discussions
# Local node package templates

#### Pre-Flight checklist

  - [x] Code is tested *Linting changes are "tested"; I have not included any tests for the templating.*

#### What are the relevant tickets?
#4010 

#### What's this PR do?
This PR adds a local node package `ol-template` that provides a template for new packages.

**Additionally:** This PR changes the linting a bit. 
- Previously, we ran linting / formatting on the `src/` directory in each package. This had two undesirable effects:
    - if there was no `src/` directory, pretter & eslint would throw an error. *That's the immediate motivation for making the change in this PR; the `ol-template` package has no `src/` directory.*
    - some configuration files (e.g., `jest.config.ts`) at package roots were not being linted / formatted.
- Now: We lint all js/ts files, except those in `build/` directories and in `node_modules/`. 



#### How should this be manually tested?

1. Enter the frontend container: `docker compose exec watch bash`.
2. Create a new local package with `yarn create-package`. After creation, you should be instructed to run `yarn install`. Go for it.
3. Add linting error (example: use `let` when `const` would do) in your local package, and save. *Depending on editor, may need to save without formatting. In VSCode, that's cmd+shift+P --> "Save without formatting".*
4. Run `yarn lint-check`. It should fail. *If the error was linting error rather than formatting. 
5. Add a formatting error (example: a semicolon) in your local package & save without formatting.
6. Run `yarn fmt-fix`. You should see the file reformatted.
7. Run `yarn workspace infinite-corridor add <your package name>`.
8. Try importing your package in `infinite-corridor`. It should work. 

#### Any background context you want to provide?

This uses [`plop`](https://www.npmjs.com/package/plop) for the templating.

I considered using [copier](https://github.com/mitodl/hq/discussions/916) for this. I did experiment with Copier, but decided against using it for this because:

- the use case here is pretty simple; plop is good enough
- it seemed more natural to create frontend packages via the `watch` container, which does not have Python.
- Copier's main distinction is its ability to update files when the template and the copy are tracked in git[^1]. This doesn't seem super useful for the local node packages because:
    
    the template has 4 files (package.json, tsconfig, jest config, and index.ts). Updates to these files packages generally happen via updating the base tsconfig or jest config, OR by installing new dependencies via `yarn`. In the case of installing new dependencies via `yarn`, the package.json ended up getting git conflicts with Copier's template.  

The templates are pretty simple, so if we end up adopting copier and want to use it here, it would be easy to convert them.

[^1]: For example, Copier seems like it would be nice for setting up common Github Actions CI files, then when breaking changes happen to, say, `actions/setup-node` or `actions/setup-python`, we could update all our repos.
